src/compiler/checker.ts

```diff
@@ -25284,6 +25284,55 @@ namespace ts {
                         }
                     }
                 }
+                if (type && type!==anyType && isPropertyAccessExpression(callExpression.expression)){

```

Comment:
About these guard conditions
- `isPropertyAccessExpression` : It is here because it matches the scope of the posted bug.  I haven't tested whether it is required.  
- `type !==anyType` : When this guard is not here the tests hang on `tests/cases/conformance/fixSignatureCaching.ts`.  Presumably this fix code  should be able to execute and successfully narrow without hanging, even when `type !==anyType`.  I this something that should be resolved before this fix can be called complete?  @ahejlsberg  - What do you think?